✨ Neutral Atom QDMI Device

ystade opened the pull request:
## Description

This PR adds a universal QDMI device implementation for neutral atom-based quantum computers. The device itself is specified in a JSON file whose structure is given by a protobuf definition. During compilation time, the JSON file is read and translated to C++ code similar to how TableGen works in the LLVM project.

Subsequent PRs will add a QDMI driver utilizing the new device and a FoMaC tailored to those neutral atom devices.

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes or removals.
- [ ] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.
- [x] As mentioned in one thread, we want to open an issue for 3D support on NA devices, i.e., including the z-variable.

ystade commented:
@burgholzer This PR is far from being ready, that's why it is labeled as `Draft`. However, it would be very helpful, if you could take a look at how the addition is structured and give feedback on this.

For context: The protobuf file in `protocol/na/` is consumed by an executable defined in `na/device/App.cpp`. This executable parses the JSON in `json/na/` and generates a C++ header files that is included in the `na/device/Device.cpp` file.

ystade commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,158 @@
+               EXPORT_NAME CoreNaDeviceGen)
+  # place in the bin directory
+  set_target_properties(${TARGET_NAME} PROPERTIES RUNTIME_OUTPUT_DIRECTORY
+                                                  "${CMAKE_BINARY_DIR}/bin")
```
I roughly remembered that qcec puts an executable in a top-level `bin` directory as well. Hence, I wanted to resemble this structure here as well. However, for qcec, the `bin` directory already exists on the project level not only on the build level. The `bin` directory is in no circumstances needed. I am happy to change this to whatever you wish for. I just did not know what to do, since it is the first (?) executable that we build in core, or at least the first that we also use than during the build process in core.

burgholzer replied:
This is definitely the first C++ executable to be included in MQT Core.
DDSIM has a couple. None of the other MQT repositories have them.
I suppose the solution here works just fine, I was just trying to understand the rationale a little more. Thanks.

ystade commented on test/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+# Licensed under the MIT License
+
+# Set test target name
+set(TARGET_NAME mqt-core-na-device-test)
```
Iirc, for some reason `MQT_CORE_TARGET_NAME` is defined in the `src/CMakeLists.txt`, hence, it is only available in the `src` subdirectories and not in the `test` subdirectories and I cannot access it here.

burgholzer replied:
This has changed with the restructuring of the Python bindings and is now in the top-level CMake file https://github.com/munich-quantum-toolkit/core/blob/a06d0fdfd1baaea82a313c1e184249a0480570d2/CMakeLists.txt#L92

ystade commented on proto/na/device.proto:
Mh, good question. The user in the end needs to provide the JSON file. Hence, I would suggest that the JSON input file needs some proper documentation of all options. For that I do not really see an other option then putting this into the docs in some Markdown file. With that said, does the protobuf still require documentation or how would you imagine the entire documentation of the format?

burgholzer replied:
Hm. I am not 100% sure. The protobuf spec is the schema for the JSON file to be put in (to some degree). I was thinking whether we could be autogenerating some documentation if we put it into the protobuf files. But that's probably overkill. Let's just go with a Markdown documentation page on this for now.

ystade commented on proto/na/device.proto:
```diff
@@ -0,0 +1,115 @@
+  }
+  repeated LocalMultiQubitOperation local_multi_qubit_operations = 8;
+
+  message ShuttlingUnit {
+    string name = 1;
+    Region region = 2;
+    uint64 num_rows = 3;
+    uint64 num_columns = 4;
+    double moving_speed = 5;
+    uint64 load_duration = 6;
+    uint64 store_duration = 7;
+    double load_fidelity = 8;
+    double store_fidelity = 9;
+    uint64 num_parameters = 10;
+  }
+  repeated ShuttlingUnit shuttling_units = 9;
```
I see your point but this would duplicate the attributes `region`, `num_rows`, and `num_columns`. The resulting redundancy could even lead to a mismatch of those numbers between corresponding load, store and move operations. So, I do not know a better solution as of now.

ystade commented on src/na/device/Generator.cpp:
```diff
@@ -0,0 +1,359 @@
+ * @returns a factor every time value must be multiplied with to convert it to
+ * microseconds.
+ */
+[[nodiscard]] auto getTimeUnit(const Device& device) -> double {
+  if (device.time_unit().unit() == "us") {
+    return static_cast<double>(device.time_unit().value());
+  }
+  if (device.time_unit().unit() == "ns") {
+    return static_cast<double>(device.time_unit().value()) * 1e-3;
+  }
+  std::stringstream ss;
+  ss << "Unsupported time unit: " << device.time_unit().unit();
+  throw std::runtime_error(ss.str());
+}
```
Since it is _our_ device, I would add them here on demand. All the code written must also be tested. Hence, I would limit it to what we actually need.

burgholzer replied:
Good point. Agreed!

ystade commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,449 @@
+ * MQT_NA_QDMI_Device_Session_impl_d.
+ */
+[[nodiscard]] auto sessions()
+    -> std::unordered_map<MQT_NA_QDMI_Device_Session,
+                          std::unique_ptr<MQT_NA_QDMI_Device_Session_impl_d>>& {
+  static std::unordered_map<MQT_NA_QDMI_Device_Session,
+                            std::unique_ptr<MQT_NA_QDMI_Device_Session_impl_d>>
```
Yes, this is quite a neat trick in my opinion. `clang-tidy` or whoever always complains about `new` and `delete` and that one should use owning memory. With this map one can use `unique_ptr` and is able to completely avoid `new` and `delete`. The only issue that must be be solved is that the interface works on the pointer not on the `unique_ptr`s. Hence, one needs this map to map the pointer to the unique_ptr. I do see the extra level of indirection here. If you prefer the `new` and `delete` with respective `NOLINT` statements, I am also fine with that.

burgholzer replied:
I think I see the value of this. However, it definitely adds an additional layer of complexity that we would not necessarily have with `new` and `delete`.
I'd say let's keep this additional complexity for now to get the ease of use and memory safety until this turns out to be a performance problem.

burgholzer commented on CMakeLists.txt:
```diff
@@ -14,6 +14,9 @@ project(
   LANGUAGES C CXX
   DESCRIPTION "MQT Core - The Backbone of the Munich Quantum Toolkit")
 
+# todo: this seems wrong here but needed to compile na qdmi targets.
+set(CMAKE_CXX_STANDARD 17)
+
```
No. We should be specifying this via
```cmake
target_compile_features(${TARGET_NAME} PRIVATE cxx_std_17)
```
For all the QDMI activities here that should totally be feasible and is actually how we handle this in MQT Core at the moment (at least before the MLIR stuff), see https://github.com/munich-quantum-toolkit/core/blob/a06d0fdfd1baaea82a313c1e184249a0480570d2/src/ir/CMakeLists.txt#L30

burgholzer commented on cmake/ExternalDependencies.cmake:
```diff
@@ -71,5 +71,39 @@ if(BUILD_MQT_CORE_TESTS)
   list(APPEND FETCH_PACKAGES googletest)
 endif()
 
+set(Protobuf_VERSION
+    31.1
+    CACHE STRING "protobuf version")
+set(Protobuf_URL
+    https://github.com/protocolbuffers/protobuf/releases/download/v${Protobuf_VERSION}/protobuf-${Protobuf_VERSION}.tar.gz
+)
+FetchContent_Declare(protobuf URL ${Protobuf_URL} FIND_PACKAGE_ARGS ${Protobuf_VERSION} CONFIG
+                                  NAMES protobuf)
+list(APPEND FETCH_PACKAGES protobuf)
+
+# cmake-format: off
+set(QDMI_VERSION 1.1.0
+        CACHE STRING "QDMI version")
+set(QDMI_REV "v${QDMI_VERSION}"
+        CACHE STRING "QDMI identifier (tag, branch or commit hash)")
+set(QDMI_REPO_OWNER "Munich-Quantum-Software-Stack"
+        CACHE STRING "QDMI repository owner (change when using a fork)")
+# cmake-format: on
+FetchContent_Declare(
+  qdmi
+  GIT_REPOSITORY https://github.com/${QDMI_REPO_OWNER}/qdmi.git
+  GIT_TAG ${QDMI_REV}
+  FIND_PACKAGE_ARGS ${QDMI_VERSION})
+list(APPEND FETCH_PACKAGES qdmi)
+
+set(SPDLOG_VERSION
+    1.15.3
+    CACHE STRING "spdlog version")
+set(SPDLOG_URL https://github.com/gabime/spdlog/archive/refs/tags/v${SPDLOG_VERSION}.tar.gz)
+# Add position independent code for spdlog, this is required for python bindings on linux
+set(SPDLOG_BUILD_PIC ON)
+FetchContent_Declare(spdlog URL ${SPDLOG_URL} FIND_PACKAGE_ARGS ${SPDLOG_VERSION})
+list(APPEND FETCH_PACKAGES spdlog)
+
```
No, you are absolutely right. I think this was just me still being anxious about every dependency we add to our projects.

burgholzer commented on src/na/device/Generator.cpp:
```diff
@@ -0,0 +1,359 @@
+
+#include "na/device/Generator.hpp"
+
+#include "na/device/device.pb.h"
```
Then the corresponding include in the `Generator.hpp` needs to be updated, which still uses `<...>` includes for this header.

burgholzer commented on src/na/device/App.cpp:
```diff
@@ -0,0 +1,146 @@
+         "Options:\n"
+         "  -h, --help          Show this help message and exit.\n"
+         "  -v, --version       Show version information and exit.\n"
+         "  -o, --output FILE   Specify the output header file. If no output\n"
+         "                      file is specified, the JSON file is just\n"
+         "                      parsed and no output produced.\n"
+         "  -s, --schema FILE   Write a JSON schema with default values.\n"
+         "                      This option does not require a JSON file.\n";
```
The first one. It would be sub-commands of the same exe

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,449 @@
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  auto uniqueSession = std::make_unique<MQT_NA_QDMI_Device_Session_impl_d>();
+  *session = sessions()
+                 .emplace(uniqueSession.get(), std::move(uniqueSession))
+                 .first->first;
```
Fine with me. I am typically not the biggest fan of these "nichtssagende" `.first->first` and am much more in favor of structured access/code.

ystade commented:
The new dependency of [`protobuf`](https://github.com/protocolbuffers/protobuf) is still giving me a few headaches. Same as `googletest` it depends on [`abseil`](https://github.com/abseil/abseil-cpp). However, when `abseil` is not installed, both fetch it but in different versions leading to two `abseil` libraries. Hence, the linker ignores the duplicate library and cannot find the abseil symbols at all during linking. This can either be fixed by installing abseil but in order not to enforce the installation of abseil to users of MQT Core, the other option is to fetch abseil centrally already in the MQT Core configuration.

When doing that, `googletest` can be instructed to use the available `abseil` by setting `GTEST_HAS_ABSL ON`. However, then `googletest` additionally requires [`re2`](https://github.com/google/re2). This then has also be added as a dependency  in the MQT Core configuration.

Now, the ugly part comes: The cmake configuration of `re2` does not allow to disable the generation of install targets. Simulteanously, the `re2`'s install targets require the export targets from `abseil`, hence, `ABSL_ENABLE_INSTALL` is set to `ON`. In summary, this adds a lot of new install targets (that are btw not associated to any components). This then causes issues when building the Python wheel because before #1021 **everything** was installed into the wheel. After #1021 is merged this should be fixed for now. 

However, when we want to ship the QDMI NA device at some point with the Python wheel, we need to be very careful what we include in the Python wheel. Because of #1021 together with the fact that the install targets of `re2` do not belong to any component, those cannot be installed anymore in this setup. However, since we should not need to have `protobuf` itself in the wheel (the auto-generated QDMI NA device should suffice), we probably do not need to install `protobuf` including the `abseil` dependency. even if we need `protobuf`, this still might be possible, as `protobuf` _does_ connect its install targets to components.

burgholzer commented on src/na/device/App.cpp:
```diff
@@ -0,0 +1,385 @@
+
+#include "na/device/Generator.hpp"
+#include "na/device/device.pb.h"
+#include "spdlog/spdlog.h"
```
```suggestion
#include <spdlog/spdlog.h>
```

burgholzer commented on src/na/device/App.cpp:
```diff
@@ -0,0 +1,385 @@
+ */
+auto printUsage(const std::string& programName) -> void {
+  std::cout
+      << "Generator of a header file from a JSON file for neutral atom QDMI\n"
+         "implementation.\n"
```
```suggestion
      << "Generator for turning neutral atom computer JSON specifications into header files to be used as part of a neutral atom QDMI device implementation.\n"
```
a bit verbose, but maybe slightly better.

burgholzer commented on src/na/device/App.cpp:
```diff
@@ -0,0 +1,396 @@
+ * @param argc is the number of command line arguments.
+ * @param argv is the array of command line arguments.
+ */
+int main(int argc, char* argv[]) {
```
This suggestion isn't terribly bad. Might as well adopt it to silence the CodeQL warning.

burgholzer commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,181 @@
+           ${MQT_CORE_INCLUDE_BUILD_DIR}/na/device/Generator.hpp
+           ${PROTO_HDR})
+
+  # Link Protobuf, QDMI, spdlog, and project warnings
```
```suggestion
  # Link Protobuf, spdlog, and project warnings
```
I don't see QDMI being linked here.

burgholzer commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,181 @@
+  # Make version available
+  target_compile_definitions(${TARGET_NAME} PRIVATE MQT_CORE_VERSION="${MQT_CORE_VERSION}")
+
+  # add to list of MQT core targets
+  set(MQT_CORE_TARGETS
+      ${MQT_CORE_TARGETS} ${TARGET_NAME}
+      PARENT_SCOPE)
```
Hm. Just a thought: do we really want to have this target and also the one below as part of the MQT_CORE_TARGETS that get installed per default with every mqt-core distribution (=Python package)?
As of now, they are merely used to generate the example device, which should definitely be part of the set of targets.
I have no clear opinion here (yet), but maybe you have.

burgholzer commented on test/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+    ${TARGET_NAME}
+    PRIVATE EXECUTABLE_PATH="$<TARGET_FILE:MQT::CoreQdmiNaDeviceGenerator>"
+            MQT_CORE_VERSION="${MQT_CORE_VERSION}"
+            NA_DEVICE_JSON="${CMAKE_SOURCE_DIR}/json/na/device.json")
```
I am not quite sure that this will work if MQT Core is used in a top-level project.
I suppose this should be something involving `CMAKE_PROJECT_DIR` or whatever the respective variable is called.

burgholzer commented on src/na/device/Generator.cpp:
```diff
@@ -0,0 +1,410 @@
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/message.h>
+#include <google/protobuf/util/json_util.h>
+#include <istream>
+#include <spdlog/spdlog.h>
```
```suggestion
#include <istream>
#include <ostream>
#include <spdlog/spdlog.h>
```

burgholzer commented on src/na/device/Generator.cpp:
```diff
@@ -0,0 +1,410 @@
+auto writeJSONSchema(const std::string& path) -> void {
+  // Write to file
+  std::ofstream ofs(path);
+  if (ofs.is_open()) {
+    writeJSONSchema(ofs);
+    ofs.close();
+    SPDLOG_INFO("JSON template written to {}", path);
+  } else {
+    std::stringstream ss;
+    ss << "Failed to open file for writing: " << path;
+    throw std::runtime_error(ss.str());
+  }
```
Truly minor and definitely a stylistic choice, but I would prefer the pattern
```c++
  std::ofstream ofs(path);
  if (!ofs.is_good()) {
    ... error ...
  }
  ... proceed with file ...
```
similar to how it is used in the read function down below.

burgholzer commented on test/na/device/test_device.cpp:
```diff
@@ -0,0 +1,510 @@
+namespace testing {
+namespace {
+// NOLINTBEGIN(readability-identifier-naming,cppcoreguidelines-avoid-const-or-ref-data-members)
+MATCHER_P2(IsBetween, a, b,
+           absl::StrCat(negation ? "isn't" : "is", " between ",
+                        PrintToString(a), " and ", PrintToString(b))) {
+  return a <= arg && arg <= b;
+}
```
is there any particular reason we are using abseil here?
If possible, I'd like to avoid using dependencies we don't actually need.
This seems like a test matcher that might already be available in google test or google mock.

ystade replied:
I really looked in the docs and couldn't find an appropriate matcher, see also https://stackoverflow.com/questions/21972285/expect-a-value-within-a-given-range-using-google-test.

This Matcher is btw. given as an example in the documentation of gtest of how to create custom matchers. Since gtest depends on abseil, this is already part of the project and I do not see a problem with that. However, the concatenation of string can obviously also implemented differently.

ystade replied:
I like the comprehensive error messages that comes with the matcher. I modified the code such that ABSEIL is not needed. Is that fine?

burgholzer commented on test/na/device/test_device.cpp:
```diff
@@ -0,0 +1,510 @@
+      testing::AnyOf(QDMI_SUCCESS, QDMI_ERROR_NOTSUPPORTED));
+}
+
+TEST_F(QDMISpecificationTest, QuerySiteProperty) {
+  MQT_NA_QDMI_Site site = querySites(session).front();
+  EXPECT_EQ(
+      MQT_NA_QDMI_device_session_query_site_property(
+          session, nullptr, QDMI_SITE_PROPERTY_INDEX, 0, nullptr, nullptr),
+      QDMI_ERROR_INVALIDARGUMENT);
+  EXPECT_EQ(MQT_NA_QDMI_device_session_query_site_property(
+                nullptr, site, QDMI_SITE_PROPERTY_INDEX, 0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT);
+  EXPECT_EQ(MQT_NA_QDMI_device_session_query_site_property(
+                session, site, QDMI_SITE_PROPERTY_MAX, 0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT);
+  EXPECT_THAT(MQT_NA_QDMI_device_session_query_site_property(
+                  session, site, QDMI_SITE_PROPERTY_NAME, 0, nullptr, nullptr),
+              testing::AnyOf(QDMI_SUCCESS, QDMI_ERROR_NOTSUPPORTED));
+}
```
Should these tests be run on every site?

ystade replied:
No these edge cases, i.e., invlaid arguments are _not_ tested in the other version. Hence, those tests are truly complementary.

burgholzer commented on test/na/device/test_device.cpp:
```diff
@@ -0,0 +1,510 @@
+              testing::AnyOf(QDMI_SUCCESS, QDMI_ERROR_NOTSUPPORTED));
+}
+
+TEST_F(QDMISpecificationTest, QueryOperationProperty) {
+  MQT_NA_QDMI_Operation operation = queryOperations(session).front();
+  EXPECT_EQ(MQT_NA_QDMI_device_session_query_operation_property(
+                nullptr, operation, 0, nullptr, 0, nullptr,
+                QDMI_OPERATION_PROPERTY_NAME, 0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT);
+  EXPECT_EQ(MQT_NA_QDMI_device_session_query_operation_property(
+                session, operation, 0, nullptr, 0, nullptr,
+                QDMI_OPERATION_PROPERTY_MAX, 0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT);
+  EXPECT_THAT(MQT_NA_QDMI_device_session_query_operation_property(
+                  session, operation, 0, nullptr, 0, nullptr,
+                  QDMI_OPERATION_PROPERTY_QUBITSNUM, 0, nullptr, nullptr),
+              testing::AnyOf(QDMI_SUCCESS, QDMI_ERROR_NOTSUPPORTED));
+}
```
Should these tests be run for every operation?

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,465 @@
+
+namespace {
+
+/**
+ * @brief Provides access to the device name.
+ * @returns a reference to a static variables that stores the
+ * device name.
+ */
+[[nodiscard]] auto name() -> std::string& {
+  static std::string name;
+  return name;
+}
+
+/**
+ * @brief Provides access to the number of qubits in the device.
+ * @returns a reference to a static variable that stores the number of qubits.
+ */
+[[nodiscard]] auto qubitsNum() -> size_t& {
+  static size_t qubitsNum = 0;
+  return qubitsNum;
+}
+
+/**
+ * @brief Provides access to the lists of sites and operations.
+ * @returns a reference to a static vector of unique pointers to
+ * @ref MQT_NA_QDMI_Site_impl_d.
+ */
+[[nodiscard]] auto sites()
+    -> std::vector<std::unique_ptr<MQT_NA_QDMI_Site_impl_d>>& {
+  static std::vector<std::unique_ptr<MQT_NA_QDMI_Site_impl_d>> sites;
+  return sites;
+}
```
At some point we talked about turning the device into a singleton similar to how this could be done for the driver. Did you try that and it was not really worth it, or was there just not enough time? I could still see the potential benefit in doing it.

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,465 @@
+    ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_FIDELITY, double,
+                              operation->fidelity, prop, size, value, sizeRet)
+  } else {
+    ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_DURATION, double, 0.0,
+                              prop, size, value, sizeRet)
+    ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_FIDELITY, double, 1.0,
+                              prop, size, value, sizeRet)
```
Does a shuttling move really always have duration 0 and fidelity 1?
In practice the duration as well as the fidelity depends on the distance of the move, right?

burgholzer replied:
Yeah. I suppose returning `NOTSUPPORTED` is the right choice here then.
But actually for both of these properties, as both of them depend on the duration and distance of the move.
The idling decoherence time is something that the device would then have to report as well, but I think we have that tracked somewhere already.

burgholzer commented on cmake/ExternalDependencies.cmake:
I have not yet looked into the dependency situation in detail. Just leaving this comment here so that I do not forget to do that.

burgholzer replied:
Nice! 🙌🏼

burgholzer reviewed (changes requested):
Thanks for all your work on this! This is really looking great already. Most of my comments are truly minor and/or stylistic nitpicks.
Other than that, there are just some conceptual questions left.

ystade commented on docs/na_qdmi_device.md:
```diff
@@ -0,0 +1,115 @@
+
+##### `traps` _(array of objects)_:
+
+- **`latticeOrigin`** _(object)_: The origin of the lattice.
+  - **`x`** _(number)_: X-coordinate of the origin.
+  - **`y`** _(number)_: Y-coordinate of the origin.
+- **`latticeVector1`** _(object)_: Defines one lattice vector.
+  - **`vector`** _(object)_: A vector in the lattice.
+    - **`x`** _(number)_: X-component of the vector.
+    - **`y`** _(number)_: Y-component of the vector.
+  - **`repeat`** _(integer)_: Number of repetitions of the vector.
+- **`latticeVector2`** _(object)_: Defines the other lattice vector.
+  - **`vector`** _(object)_: A vector in the lattice.
+    - **`x`** _(number)_: X-component of the vector.
+    - **`y`** _(number)_: Y-component of the vector.
+  - **`repeat`** _(integer)_: Number of repetitions of the vector.
+- **`sublatticeOffsets`** _(array of objects)_: Offsets for sublattices.
+  - **`x`** _(number)_: X-offset.
+  - **`y`** _(number)_: Y-offset.
```
I would like to postpone this addition because right now I see no value in adding it. I add it as one todo item and open the respective issue after this PR is complete.

burgholzer replied:
Works for me 👍🏼

ystade commented on docs/na_qdmi_device.md:
```diff
@@ -0,0 +1,115 @@
+
+##### `lengthUnit` and `timeUnit` _(objects)_:
+
+- **`value`** _(number)_: The scaling factor for the unit.
+- **`unit`** _(string)_: The unit of measurement (e.g., "um" for micrometers, "us" for microseconds).
```
I am uncertain whether it is required but I though when the grid size is for example 4µm, then you can specify all distances in multiples of 4µm. In this scenario, the unit would be "um" and the value would be 4. However, I am not sure whether I actually understood your question correctly.

burgholzer replied:
Ahhh. Yeah, you understood the question correctly and now I also understand the intention behind the factor. I agree it provides value. I'd still argue that it could potentially use a better name (like `scaling_factor`).

burgholzer commented on docs/na_qdmi_device.md:
```diff
@@ -0,0 +1,115 @@
+#### Top-Level Fields
+
+- **`name`** _(string)_: The name of the device.
+- **`num_qubits`** _(integer)_: The total number of qubits in the device.
```
All options below use camelCase, at least in the documentation page here, which is the main reason why I annotated this as a comment. I don't particularly care about the casing as long as it is consistent across the protofile definition.

ystade commented:
@burgholzer the two main major changes after your last review are:
1. the device is implented as a singleton, and
2. protobuf (incl. abseil) is removed as dependency and replaced with nlohmann_json and a corresponding `struct` definition.

I have now one major issue with the tests: Since the device does not support the job interface, one is not able to create a job pointer that is `!= nullptr`. Hence, all functions in the job interface will immediately return `INVALIDARGUMENT` and right now I do not see how I can test the code of the job interface. Any ideas? Perhaps, removing the mock implementation of the job interface entirely?

burgholzer commented:
> I have now one major issue with the tests: Since the device does not support the job interface, one is not able to create a job pointer that is `!= nullptr`. Hence, all functions in the job interface will immediately return `INVALIDARGUMENT` and right now I do not see how I can test the code of the job interface. Any ideas? Perhaps, removing the mock implementation of the job interface entirely?

Just quickly commenting on this. Haven't had time to look through the rest yet.
How come that you can't implement a job pointer that is not nullptr?
Shouldn't the job creation function be responsible to populate a job? (Even if it's just a dummy job that merely tracks the session)
I think the job interface needs to be implemented at least to that degree that one can create a job that is not null.

ystade commented:
> > I have now one major issue with the tests: Since the device does not support the job interface, one is not able to create a job pointer that is `!= nullptr`. Hence, all functions in the job interface will immediately return `INVALIDARGUMENT` and right now I do not see how I can test the code of the job interface. Any ideas? Perhaps, removing the mock implementation of the job interface entirely?
> 
> Just quickly commenting on this. Haven't had time to look through the rest yet. How come that you can't implement a job pointer that is not nullptr? Shouldn't the job creation function be responsible to populate a job? (Even if it's just a dummy job that merely tracks the session) I think the job interface needs to be implemented at least to that degree that one can create a job that is not null.

@burgholzer Now, the test coverage is at 97%. The PR is now ready for a review again. From my side only the missing properties in QDMI are missing and the PR here itself is complete.

burgholzer commented on src/na/device/App.cpp:
```diff
@@ -0,0 +1,404 @@
+    return 1;
+  }
+  // unpack the parsed arguments and the index of the first sub-command here
+  // because strcutured bindings only work with fresh variables
```
```suggestion
  // because structured bindings only work with fresh variables
```

burgholzer commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,148 @@
+               EXPORT_NAME CoreNaDeviceGen)
+
+  # set c++ standard
+  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_17)
```
```suggestion
  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_20)
```

burgholzer commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,148 @@
+                                                  "${CMAKE_BINARY_DIR}/bin")
+
+  # set c++ standard
+  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_17)
```
```suggestion
  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_20)
```

burgholzer commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,148 @@
+  target_compile_definitions(${TARGET_NAME} PRIVATE MQT_CORE_VERSION="${MQT_CORE_VERSION}")
+
+  # Create an alias for the target
+  add_executable(MQT::CoreQdmiNaDeviceGenerator ALIAS ${TARGET_NAME})
```
Should. We always capitalize QDMI?
Also, none of the other target names or aliases above contain QDMI. Maybe all of them should.

burgholzer commented on src/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,148 @@
+    PRIVATE MQT::ProjectOptions MQT::ProjectWarnings spdlog::spdlog)
+
+  # set c++ standard
+  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_17)
```
```suggestion
  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_20)
```
This should really be moved into the `add_mqt_library` macro.

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,499 @@
+
+#include "na/device/Device.hpp"
+
+#include "Initializer.hpp"
```
This header confuses me. Can we give it a slightly more appropriate name? Could we perhaps namespace it properly?

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,499 @@
+  INITIALIZE_OPERATIONS(operations);
+  INITIALIZE_SITES(sites);
+}
+Device::~Device() { sessions.clear(); }
```
Is this even necessary? The default constructor will delete the sessions anyway, right?

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,499 @@
+  if (session == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  auto uniqueSession = std::make_unique<MQT_NA_QDMI_Device_Session_impl_d>();
+  *session = sessions.emplace(uniqueSession.get(), std::move(uniqueSession))
+                 .first->first;
+  return QDMI_SUCCESS;
```
I still find this pattern utterly confusing; especially the `.first->first` part makes this increasingly hard to read in my opinion.
I'd much prefer something like
```suggestion
  auto uniqueSession = std::make_unique<MQT_NA_QDMI_Device_Session_impl_d>();
  auto [it, inserted] = sessions.emplace(uniqueSession.get(), std::move(uniqueSession));
  *session = it->first;
  return QDMI_SUCCESS;
```

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,499 @@
+  return QDMI_SUCCESS;
+}
+auto MQT_NA_QDMI_Device_Session_impl_d::freeDeviceJob(
+    [[maybe_unused]] MQT_NA_QDMI_Device_Job job) -> void {
```
```suggestion
    MQT_NA_QDMI_Device_Job job) -> void {
```

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,499 @@
+                                  size, value, sizeRet);
+}
+auto MQT_NA_QDMI_Device_Job_impl_d::free() -> void {
+  // The called function is only static because jobs are not supported. We keep
+  // the function call, however, as it is needed to free the job when the jobs
+  // are supported.
+  //===--------------------------------------------------------------------===//
+  // NOLINTNEXTLINE(readability-static-accessed-through-instance)
```
The function is no longer static
```suggestion
```

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,499 @@
+  if (job == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  // The called function is only static because jobs are not supported. We keep
+  // the function call, however, as it is needed to free the job when the jobs
+  // are supported.
```
This is just copy-pasted and needs to be adapted for the respective function. Also holds below.

burgholzer commented on src/na/device/Generator.cpp:
```diff
@@ -0,0 +1,380 @@
+#include <cstdint>
+#include <fstream>
+#include <istream>
+#include <nlohmann/detail/exceptions.hpp>
```
As in the other file, the details headers should not be included (they also do not exist when using the single-header version of nlohman_json).

burgholzer commented on src/na/device/Generator.cpp:
```diff
@@ -0,0 +1,380 @@
+                     std::ostream& os) -> void {
+  os << "#define INITIALIZE_OPERATIONS(var) var.clear();";
+  for (const auto& operation : device.globalSingleQubitOperations) {
+    os << "\\\n"
+          "  var.emplace_back(std::make_unique<MQT_NA_QDMI_Operation_impl_d>("
+          "MQT_NA_QDMI_Operation_impl_d{\""
+       << operation.name
+       << "\", MQT_NA_QDMI_Operation_impl_d::Type::GlobalSingleQubit, "
+       << operation.numParameters << ", 1, "
+       << static_cast<double>(operation.duration) * timeUnit << ", "
+       << operation.fidelity << "}));";
```
I think the patterns throughout this file can be simplified a little bit because `std::make_unique` forwards arguments. Something like the following should work
```suggestion
    os << "\\\n"
          "  var.emplace_back(std::make_unique<MQT_NA_QDMI_Operation_impl_d>(\""
       << operation.name
       << "\", MQT_NA_QDMI_Operation_impl_d::Type::GlobalSingleQubit, "
       << operation.numParameters << ", 1, "
       << static_cast<double>(operation.duration) * timeUnit << ", "
       << operation.fidelity << "));";
```
It might be that even the `std::make_unique` is superfluous

burgholzer commented on test/na/device/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+            MQT_CORE_VERSION="${MQT_CORE_VERSION}"
+            NA_DEVICE_JSON="${PROJECT_SOURCE_DIR}/json/na/device.json")
+  # Set c++ standard
+  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_17)
```
```suggestion
  target_compile_features(${TARGET_NAME} PRIVATE cxx_std_20)
```

burgholzer commented on docs/index.md:
```diff
@@ -39,6 +39,7 @@ mqt_core_ir
 dd_package
 zx_package
 mlir
+NA QDMI Device <na_qdmi_device>
```
Hm. I see the point. However, this makes it somewhat inconsistent with the remaining docs sections (not only in code, but also in the real ToC).
Would an abbreviation for the other sections make sense as well?

burgholzer replied:
Makes sense. I must have mixed up the MQT Core docs with some other MQT docs, where this "no line break" was definitely not the case.
I am on board 🫡 👍🏼

ystade commented on docs/na_qdmi_device.md:
```diff
@@ -0,0 +1,121 @@
+## Describing a Device
+
+The basis of a such device implementation is a specification in a JSON file.
+The structure of this JSON files is defined via `struct`s in [`Generator.hpp`](../include/mqt-core/na/device/Generator.hpp) that define functions to serialize and deserialize the data using the [nlohmann/json](https://json.nlohmann.me) library.
+During compilation, this JSON file is parsed and the corresponding C++ code is produced by an application (see [`App.cpp`](../src/na/device/App.cpp)) for the actual QDMI device implementation.
```
Can you give me a hint, how I can reference C++ symbols from that part (which afaik goes through myst)?

burgholzer replied:
See https://breathe.readthedocs.io/en/latest/domains.html
The respective symbols etc. should be available under the `cpp:` domain.

burgholzer replied:
I will take a look after my meeting! 👍🏼

burgholzer replied:
See https://github.com/munich-quantum-toolkit/core/pull/996/commits/2a43ce55a9bc22ba0e760377a86a492bd8c695f0 for how to create a proper cross reference to the C++ code

ystade replied:
Ah, nice, thanks!

ystade commented on include/mqt-core/na/device/Device.hpp:
```diff
@@ -0,0 +1,314 @@
+  explicit MQT_NA_QDMI_Device_Job_impl_d(
+      MQT_NA_QDMI_Device_Session_impl_d* session)
+      : session(session) {}
+  /**
+   * @brief Frees the device job.
+   * @note This function just forwards to the session's @ref freeDeviceJob
+   * function. This function is needed because the interface only provides the
+   * job handle to the @ref QDMI_job_free function and the job's session handle
+   * is private.
+   * @see QDMI_job_free
+   */
+  auto free() -> void;
```
No it can't in this case. I think the docstring provides the respective answer. This exact function here is needed for forwarding purposes that will eventually call the function `erase` on some map calling the actual destructor of the job.

ystade commented on include/mqt-core/na/device/Generator.hpp:
Yes, we should definitely do that. Again, can I reference the C++ docu from a markdown file in the docs?

burgholzer commented on include/mqt-core/na/device/Device.hpp:
```diff
@@ -0,0 +1,314 @@
+   */
+  auto sessionFree(MQT_NA_QDMI_Device_Session session) -> void;
+
+  /**
+   * @brief Query a device property.
+   * @see MQT_NA_QDMI_device_session_query_device_property
+   */
+  auto queryProperty(QDMI_Device_Property prop, size_t size, void* value,
+                     size_t* sizeRet) -> int;
```
Fair enough. I am convinced 👍🏼

burgholzer commented:
> > > @burgholzer Now the Linter job passes and CodeQL fails due to an actual warning. However, this function does exist on purpose, should we just ignore the warning?
> 
> > 
> 
> > Is there anything reasonable we could be doing in the `finalize` method so that the warning would go away? Otherwise I am also happy to dismiss the warning here.
> 
> 
> 
> Usually, one would free all memory that was allocated for attributes and members of the device. However, since we have implemented the device as a singleton in C++ and rely on the destructors for all memory deallocations there is actually nothing reasonable to do that I can think of. 

In that case, let's just ignore the warning as "won't fix".

burgholzer commented on CHANGELOG.md:
```diff
@@ -9,6 +9,10 @@ This project adheres to [Semantic Versioning], with the exception that minor rel
 
+### Added
+
+- ✨ Add MQT's implementation of a QDMI Device for neutral atom-based quantum computing ([#996]) ([**@li-mingbao**])
```
```suggestion
- ✨ Add MQT's implementation of a QDMI Device for neutral atom-based quantum computing ([#996]) ([**@ystade**])
```
Pretty sure this is what you wanted to write here ;o)

ystade commented:
The support for all new features that were recently added to QDMI lead to way more changes and code then I expected. Especially, the tests that I added to test the new features comprise quite some lines of code. At least, more information from the JSON specification of the device than before is now actually available through the QDMI device.

ystade commented:
We decided to split the PR into two. Hence, I removed the last couple of commits that added the functionality to query operations for the supported sites. Those additions are now moved to #1100. Consequently, this PR should now be ready to be merged. 🚀

burgholzer commented on src/na/device/Device.cpp:
```diff
@@ -0,0 +1,491 @@
+#include "na/device/Device.hpp"
+
+#include "mqt_na_qdmi/device.h"
+#include "na/device/DeviceMemberInitializers.hpp"
```
Feel free to simply resolve this, but given that the header is placed in `na/device` already, I would have named it `MemberInitializers.hpp`.

burgholzer reviewed (approved):
This looks wonderful. I appreciate all the effort you put into this!
It's ready to merge now. Feel free to decide on your own whether you like the one inline suggestion. Otherwise simply resolve the comment and merge the PR 😌
